File: cmpF.m
rng(8);

if(1)
    nz = 32;
    eh = 1;
    
    %gfn = @(t,s) exp(i*2*pi*t*transpose(s));
    gfn = @(t,s) (s.').^t;
    
    %ng = 32;    gs = exp(2*pi*i*[1/2:ng]'/ng);         T = gfn(zs,gs);
    ng = 32;    tmp = [1/2:ng]'/ng - 1/2;    gs = exp(2*pi*i*tmp);
end

for it=1:3
    if(it==1)
        zs = [0:nz-1]'*eh;
    elseif(it==2)
        zs = zs + randn(size(zs))*0.1;
    else
        zs = sort(rand(nz,1)*nz);    
    end
    T = gfn(zs,gs);    for g=1:size(T,2);        T(:,g) = T(:,g)/norm(T(:,g));    end
    M = T * diag( exp(2*pi*i*tmp*eh) )*pinv(T,norm(T,'fro')*1e-2);
    
    FS=18;
    figure; imagesc(abs(M)); colorbar; axis equal; axis tight;
    set(gca, 'FontSize', FS);    bb=get(gca);    set(bb.XLabel, 'FontSize', FS);    set(bb.YLabel, 'FontSize', FS);    set(bb.ZLabel, 'FontSize', FS);    set(bb.Title, 'FontSize', FS);
    filename = sprintf('cmp_%i',it);
    print(gcf, '-depsc', filename);
end

File: python/cmpF.py
from __future__ import annotations

import numpy as np

from mt_utils import normalize_columns, pinv_abs, save_eps, set_matplotlib_headless


def main() -> None:
    set_matplotlib_headless()
    import matplotlib.pyplot as plt

    rng = np.random.default_rng(8)

    nz = 32
    eh = 1.0

    # MATLAB: tmp = [1/2:ng]'/ng - 1/2; gs = exp(2*pi*i*tmp);
    ng = 32
    tmp = (np.arange(1, ng + 1, dtype=float) / ng) - 0.5
    gs = np.exp(2j * np.pi * tmp)

    def gfn(t: np.ndarray, s: np.ndarray) -> np.ndarray:
        # MATLAB: (s.').^t
        return (s[None, :] ** t[:, None]).astype(complex)

    # iterate over three different z samplings
    zs = None
    for it in range(1, 4):
        if it == 1:
            zs = (np.arange(nz, dtype=float) * eh).reshape(-1, 1)
        elif it == 2:
            assert zs is not None
            zs = zs + rng.standard_normal(size=zs.shape) * 0.1
        else:
            zs = np.sort(rng.random((nz, 1)) * nz, axis=0)

        T = normalize_columns(gfn(zs[:, 0], gs))
        D = np.diag(np.exp(2j * np.pi * tmp * eh))
        tol = np.linalg.norm(T, ord="fro") * 1e-2
        M = T @ D @ pinv_abs(T, tol)

        plt.figure()
        plt.imshow(np.abs(M), aspect="equal", interpolation="nearest")
        plt.colorbar()
        plt.axis("tight")
        plt.title(f"cmpF it={it}")

        save_eps(f"cmp_{it}")
        plt.close()


if __name__ == "__main__":
    main()

